src/TraceEvent/BPerf/BPerfEventSource.cs

```diff
@@ -0,0 +1,312 @@
+
+                        eventRecord->UserData = new IntPtr(uncompressedBufferPtr + bufferOffset);
+
+                        bufferOffset += eventRecord->UserDataLength;
+                        bufferOffset = AlignUp(bufferOffset, 8);
+
+                        eventRecord->ExtendedData = eventRecord->ExtendedDataCount > 0 ? (TraceEventNativeMethods.EVENT_HEADER_EXTENDED_DATA_ITEM*)(uncompressedBufferPtr + bufferOffset) : (TraceEventNativeMethods.EVENT_HEADER_EXTENDED_DATA_ITEM*)0;
+
+                        bufferOffset += sizeof(TraceEventNativeMethods.EVENT_HEADER_EXTENDED_DATA_ITEM) * eventRecord->ExtendedDataCount;
+
+                        for (ushort i = 0; i < eventRecord->ExtendedDataCount; ++i)
+                        {
+                            eventRecord->ExtendedData[i].DataPtr = (ulong)(uncompressedBufferPtr + bufferOffset);
+
+                            bufferOffset += eventRecord->ExtendedData[i].DataSize;
+                            bufferOffset = AlignUp(bufferOffset, 8);
+                        }
+
+                        bufferOffset = AlignUp(bufferOffset, 16);
+
+                        if (this.sessionStartTimeQPC == 0)
+                        {
+                            if (!this.Initialize(eventRecord))
+                            {
+                                continue;
+                            }
+                            else
+                            {
+                                return -1; // -1 means initialization succeeded.
+                            }
+                        }
+
+                        // BPerf puts 65535 for Classic Events, TraceEvent fires an assert for that case.
+                        if ((eventRecord->EventHeader.Flags & TraceEventNativeMethods.EVENT_HEADER_FLAG_CLASSIC_HEADER) != 0)
+                        {
+                            eventRecord->EventHeader.Id = 0;
+                        }
+
+                        var traceEvent = this.Lookup(eventRecord);
+                        traceEvent.DebugValidate();
+
+                        if (traceEvent.NeedsFixup)
+                        {
+                            traceEvent.FixupData();
+                        }
+
+                        this.Dispatch(traceEvent);
+                        this.sessionEndTimeQPC = eventRecord->EventHeader.TimeStamp;
+                    }
+
+                    offset += compressedBufferSize;
+                    offset = AlignUp(offset, 4);
+                }
+            }
+
+            return offset;
+        }
+
+        private static int AlignUp(int num, int align)
+        {
+            return (num + (align - 1)) & ~(align - 1);
+        }
+
+        [DllImport("ntdll.dll")]
+        private static extern unsafe uint RtlDecompressBufferEx(ushort compressionFormat, byte* uncompressedBuffer, int uncompressedBufferSize, byte* compressedBuffer, int compressedBufferSize, out int finalUncompressedSize, byte* workSpace);
```

Comment:
This means that this source will only work on Windows.  I take it that is OK?

Reply:
Yes, this is Windows only.